tianshou/policy/modelbase/psrl.py

```diff
@@ -0,0 +1,197 @@
+        if self.updated is False:
+            self.solve_policy()
+        return self.policy[obs]
+
+
+class PSRLPolicy(BasePolicy):
+    """Implementation of Posterior Sampling Reinforcement Learning.
+
+    Reference: Strens M. A Bayesian framework for reinforcement learning [C]
+    //ICML. 2000, 2000: 943-950.
+
+    :param np.ndarray p_prior: dirichlet prior (alphas).
+    :param np.ndarray rew_mean_prior: means of the normal priors of rewards.
+    :param np.ndarray rew_std_prior: standard deviations of the normal
+        priors of rewards.
+    :param float discount_factor: in [0, 1].
+
+    .. seealso::
+
+        Please refer to :class:`~tianshou.policy.BasePolicy` for more detailed
+        explanation.
+    """
+
+    def __init__(
+        self,
+        p_prior: np.ndarray,
+        rew_mean_prior: np.ndarray,
+        rew_std_prior: np.ndarray,
+        discount_factor: float = 0.0,
+        **kwargs: Any,
+    ) -> None:
+        super().__init__(**kwargs)
+        self.model = PSRLModel(p_prior, rew_mean_prior, rew_std_prior)
+        assert 0.0 <= discount_factor <= 1.0, \
+            "discount factor should in [0, 1]"
+        self._gamma = discount_factor
+        self.eps = 0.0
+
+    def set_eps(self, eps: float) -> None:
+        """Set the eps for epsilon-greedy exploration."""
+        self.eps = eps
+
+    def process_fn(
+            self, batch: Batch, buffer: ReplayBuffer, indice: np.ndarray
+    ) -> Batch:
+        r"""Compute the discounted returns for each frame:
+
+        .. math::
+
+            G_t = \sum_{i=t}^T \gamma^{i-t}r_i
+
+        , where :math:`T` is the terminal time step, :math:`\gamma` is the
+        discount factor, :math:`\gamma \in [0, 1]`.
+        """
+        return self.compute_episodic_return(
+            batch, gamma=self._gamma, gae_lambda=1.0)
+
+    def forward(
+        self,
+        batch: Batch,
+        state: Optional[Union[dict, Batch, np.ndarray]] = None,
+        eps: Optional[float] = None,
+        **kwargs: Any,
+    ) -> Batch:
+        """Compute action over the given batch data with PSRL model.
+
+        :return: A :class:`~tianshou.data.Batch` with "act" key containing
+            the action.
+
+        .. seealso::
+
+            Please refer to :meth:`~tianshou.policy.BasePolicy.forward` for
+            more detailed explanation.
+        """
+        act = self.model(batch.obs, state=state, info=batch.info)
+        if eps is None:
+            eps = self.eps
+        if not np.isclose(eps, 0):
+            for i in range(len(act)):
+                if np.random.rand() < eps:
+                    act[i] = np.random.randint(0, self.model.n_action)
+        return Batch(act=act)
+
+    def learn(  # type: ignore
+        self, batch: Batch, batch_size: int, repeat: int, **kwargs: Any
+    ) -> Dict[str, float]:
+        p = np.zeros((self.model.n_action, self.model.n_state,
+                      self.model.n_state))
```

Comment:
Same as in the observe function, "p" should be renamed as "trans_count" for clarity.